### PR TITLE
Extract IPC out of webrender

### LINKED ISSUE #13480: intermittent crash ("Connection reset by peer" from create_api) in /html/semantics/embedded-content/the-canvas-element/2d.getcontext.shared.html, /html/semantics/scripting-1/the-template-element/template-element/node-document-changes.html, /dom/traversal/TreeWalker-walking-outside-a-tree.html, /XMLHttpRequest/responseText-status.html, /_mozilla/css/line_break_nowrap.html
```
 ▶ CRASH [expected FAIL] /css21_dev/html4/vertical-align-007.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } }) (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }, at ../src/libcore/result.rs:799)
  │ stack backtrace:
  │    0:     0x7fa759924c8d - backtrace::backtrace::trace::h0e60ef08c7c34e9f
  │    1:     0x7fa759925312 - backtrace::capture::Backtrace::new::h8bf319c36d8f5d1b
  │    2:     0x7fa75846e9cd - servo::main::_{{closure}}::h911afab4cc009f33
  │    3:     0x7fa759f03745 - std::panicking::rust_panic_with_hook::hb1322e5f2588b4db
  │    4:     0x7fa759f03612 - std::panicking::begin_panic::hfbeda5aad583dc32
  │    5:     0x7fa759f03550 - std::panicking::begin_panic_fmt::h4fe9fb9d5109c4bf
  │    6:     0x7fa759f034d1 - rust_begin_unwind
  │    7:     0x7fa759f3ff5c - core::panicking::panic_fmt::h4395919ece15c671
  │    8:     0x7fa759a8964f - core::result::unwrap_failed::h3191a47cbfbe2a26
  │    9:     0x7fa759a8ad20 - webrender_traits::api::_<impl webrender_traits..RenderApiSender>::create_api::h74e2147a94b73b3e
  │   10:     0x7fa7587e10f7 - _<layout_thread..LayoutThread as layout_traits..LayoutThreadFactory>::create::_{{closure}}::hcbdf224d94fa1d3a
  │   11:     0x7fa75875c796 - std::panicking::try::do_call::h032f92c4e6f5b87e
  │   12:     0x7fa759f0c0c6 - __rust_maybe_catch_panic
  │   13:     0x7fa758783e44 - _<F as alloc..boxed..FnBox<A>>::call_box::h018a2ddf52f3d72e
  │   14:     0x7fa759f01d80 - std::sys::thread::Thread::new::thread_start::h022e3887023c6290
  │   15:     0x7fa756378183 - start_thread
  │   16:     0x7fa755e8f37c - clone
  │   17:                0x0 - <unknown>
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } })
  └ Pipeline failed in hard-fail mode.  Crashing!
```

### LINKED ISSUE #20172: Extract IPC usage from WebRender into Servo
The goal of this work is to remove all references to the [ipc-channel](https://github.com/servo/ipc-channel/) crate from [WebRender](https://github.com/servo/webrender/) and move all IPC usage into the calling code in Servo instead. This will require a few parts:
* replace all uses of IPC channels in WebRender with [std::mpsc channels](https://doc.servo.org/std/sync/mpsc/fn.channel.html) instead
* [use](https://doc.rust-lang.org/cargo/reference/source-replacement.html) the modified version of WebRender in Servo
* create an IPC channel that is stored in Constellation which is used for cross-process communication with WebRender
* create a new type like webrender_api::RenderApiSender that uses an IPC channel to pass the messages to the parent
* make the IPC channel route the messages to the in-process webrender_api::RenderApiSender

Code: `components/layout_thread/lib.rs`, `components/layout_traits/lib.rs`, `components/constellation/constellation.rs`, `components/constellation/pipeline.rs`

To consider the task complete, it should be enough to demonstrate that `./mach run tests/html/about-blank.html` still renders a page after the changes are made.

### PR DESCRIPTION
**Pull request by jdm**
Fixes #20172. Fixes #13480.

### CODE REVIEW

**Comment by jdm** [comment]
Note to self:
* add missing documentation
* fix layout_2020 build

**Comment by jdm** [comment]
r? @asajeffrey 

I realized that this can land separately from https://github.com/jdm/webrender/tree/no-ipc; all we have to do is stop using the optional `ipc` feature from WebRender.

**Comment by asajeffrey** [comment]
Ooh this looks pretty good. I'll have a proper look later today.

File: components/canvas/lib.rs [hunk 1]
```diff
@@ -22,3 +26,13 @@ pub mod canvas_paint_thread;
 mod webgl_limits;
 mod webgl_mode;
 pub mod webgl_thread;
+
+pub enum ConstellationCanvasMsg {
```
Comment (asajeffrey) [inline components/canvas/lib.rs]: If we put this message type into `canvas_traits` could we make constellation not depend on `canvas`?

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -452,7 +461,7 @@ pub struct Constellation<Message, LTF, STF> {
     webxr_registry: webxr_api::Registry,
 
     /// A channel through which messages can be sent to the canvas paint thread.
-    canvas_chan: IpcSender<CanvasMsg>,
+    canvas_chan: (Sender<ConstellationCanvasMsg>, IpcSender<CanvasMsg>),
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Split this into two fields?

File: tests/wpt/metadata/css/CSS2/tables/table-anonymous-objects-009.xht.ini [hunk 3]
```diff
@@ -1,5 +1,3 @@
 [table-anonymous-objects-009.xht]
   type: reftest
   bug: https://github.com/servo/servo/issues/21092
-  expected:
-    if os == "linux": FAIL
```
Comment (asajeffrey) [inline tests/wpt/metadata/css/CSS2/tables/table-anonymous-objects-009.xht.ini]: Delete this file? (And all the other ones just like it.)

File: tests/wpt/metadata/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/fieldset-transform-translatez.html.ini [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[fieldset-transform-translatez.html]
+  expected: FAIL
```
Comment (asajeffrey) [inline tests/wpt/metadata/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/fieldset-transform-translatez.html.ini]: Hmm, a new `FAIL`?

**Review by asajeffrey — APPROVED** [review]
LGTM, some minor nits. Lots of new tests that pass yay! One new test that fails, do you know why?

File: components/canvas/lib.rs (hunk 1 above)
Comment (jdm) [reply components/canvas/lib.rs]: Right now the constellation is responsible for starting the canvas paint thread. I can file a followup to investigate removing that, but I'd rather not delay this work for it.

File: tests/wpt/metadata/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/fieldset-transform-translatez.html.ini (hunk 4 above)
Comment (jdm) [reply tests/wpt/metadata/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/fieldset-transform-translatez.html.ini]: There are minute pixel differences in the text rendering :/
Screen Shot 2019-11-27 at 1 47 28 PM

Comment (asajeffrey) [reply tests/wpt/metadata/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/fieldset-transform-translatez.html.ini]: Oh that's annoying :/

**Comment by jdm** [comment]
Hooray, more minute pixel differences on macOS for /_mozilla/css/inline_block_baseline_a.html:
Screen Shot 2019-11-27 at 3 39 32 PM

**Pull request merged**